### PR TITLE
[v1.5.1] fix clip_grad_norm to work with parameters on the different …

### PR DESCRIPTION
**Pull request by gchanan**
…devices (#38615)

Summary:
Per title.
We move all the individual gradient norms to a single device before stacking (no-op if all the gradients are already on a single device), `clip_coef` is copied to the device of gradient, which may be suboptimal as there could be multiple copies, but no worse than when we were synchronizing for each parameter. In a simple case of all gradients on a single device, there should be no synchronization.
Also, we no longer error out if parameter list is empty or none of the parameters have gradients, and return 0 total_norm instead.
Fixes https://github.com/pytorch/pytorch/issues/38605
Pull Request resolved: https://github.com/pytorch/pytorch/pull/38615

Reviewed By: ailzhang

Differential Revision: D21634588

Pulled By: ngimel

fbshipit-source-id: ea4d08d4f3445438260052820c7ca285231a156b

### CODE REVIEW

**Comment by gchanan** [comment]
FYI @dlibenzi, since it sounded like you made some changes in response to this on your side.

**Review by ngimel — APPROVED** [review]

**Comment by dlibenzi** [comment]
> FYI @dlibenzi, since it sounded like you made some changes in response to this on your side.

We will have to update our patch to undo those changes.

**Comment by gchanan** [comment]
@dlibenzi should I wait to merge this or go ahead?

**Comment by dlibenzi** [comment]
> @dlibenzi should I wait to merge this or go ahead?

I guess I can follow up with a PR on our side soon after you merged.
LMK as soon as it goes in ...

**Comment by dlibenzi** [comment]
> > @dlibenzi should I wait to merge this or go ahead?
> 
> I guess I can follow up with a PR on our side soon after you merged.
> LMK as soon as it goes in ...

Hmm, is this already merged?

https://github.com/pytorch/pytorch/blob/c2133179a9ce0063ab507911315bc2ff1b0580ac/torch/nn/utils/clip_grad.py#L27

**Comment by ngimel** [comment]
@dlibenzi it's merged in master, but now we are cherry-picking changes to 1.5.1 branch.

**Comment by ailzhang** [comment]
@gchanan This PR should be good to go. Thanks for waiting!
